From: Sven Roederer <devel-sven@geroedel.de>
Date: Thu, 24 Oct 2019 17:46:28 +0200
Subject: tools/mkimake: fix handling of long pathnames

Current pathlength of cmd to call dtc is limited to 512 bytes. Using two 245 byte
paths for itb and the required parameters go over this limit and will be cut off.

/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/scripts/mkits.sh -D asus_map-ac2200 -o /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_map-ac2200-initramfs-fit-uImage.itb.its -k /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_map-ac2200-initramfs-fit-uImage.itb -d /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-4.14.149/arch/arm/boot/dts/qcom-ipq4019-map-ac2200.dtb -C lzma -a 0x80208000 -e 0x80208000 -c "config@1" -A arm -v 4.14.149
PATH=/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-4.14.149/scripts/dtc:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/host/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/host/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/host/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games mkimage -f /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_map-ac2200-initramfs-fit-uImage.itb.its /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_map-ac2200-initramfs-fit-uImage.itb.new
sh: 1: Syntax error: Unterminated quoted string
mkimage: Can't open /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_map-ac2200-initramfs-fit-uImage.itb.new.tmp: No such file or directory
make[6]: *** [/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_map-ac2200-initramfs-fit-uImage.itb] Error 255
/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/scripts/mkits.sh -D asus_rt-ac58u -o /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_rt-ac58u-initramfs-fit-uImage.itb.its -k /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_rt-ac58u-initramfs-fit-uImage.itb -d /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-4.14.149/arch/arm/boot/dts/qcom-ipq4018-rt-ac58u.dtb -C lzma -a 0x80208000 -e 0x80208000 -c "config@1" -A arm -v 4.14.149
PATH=/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-4.14.149/scripts/dtc:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/host/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-7.4.0_musl_eabi/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/host/bin:/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/staging_dir/host/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games mkimage -f /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_rt-ac58u-initramfs-fit-uImage.itb.its /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_rt-ac58u-initramfs-fit-uImage.itb.new
sh: 1: Syntax error: Unterminated quoted string
mkimage: Can't open /var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_rt-ac58u-initramfs-fit-uImage.itb.new.tmp: No such file or directory
make[6]: *** [/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/freifunk-berlin-dev-daily-1907-f66973b-ipq40xx-generic-asus_rt-ac58u-initramfs-fit-uImage.itb] Error 255
make[6]: Leaving directory `/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/target/linux/ipq40xx/image'
make[5]: *** [install] Error 2
make[5]: Leaving directory `/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/target/linux/ipq40xx'
make[4]: *** [install] Error 2
make[4]: Leaving directory `/var/lib/buildbot/slaves/slave/ipq40xx-generic/build/firmware/openwrt/target/linux'

diff --git a/tools/mkimage/patches/110-fix-handling-of-long-pathnames.patch b/tools/mkimage/patches/110-fix-handling-of-long-pathnames.patch
new file mode 100644
index 0000000000000000000000000000000000000000..9b7febd489c3415cccdf8e92ba8f4eede8c775a7
--- /dev/null
+++ b/tools/mkimage/patches/110-fix-handling-of-long-pathnames.patch
@@ -0,0 +1,32 @@
+diff -ur u-boot-2018.03.orig/tools/fit_image.c u-boot-2018.03/tools/fit_image.c
+--- u-boot-2018.03.orig/tools/fit_image.c	2018-03-13 13:02:19.000000000 +0100
++++ u-boot-2018.03/tools/fit_image.c	2019-10-24 17:16:23.914656651 +0200
+@@ -19,6 +19,7 @@
+ #include "mkimage.h"
+ #include <image.h>
+ #include <stdarg.h>
++#include <string.h>
+ #include <version.h>
+ #include <u-boot/crc.h>
+ 
+@@ -658,6 +659,9 @@
+ 		snprintf(cmd, sizeof(cmd), "cp \"%s\" \"%s\"",
+ 			 params->imagefile, tmpfile);
+ 	}
++	if (strlen(cmd) >= MKIMAGE_MAX_DTC_CMDLINE_LEN - 1) {
++		fprintf(stderr, "WARNING: command-line for FIT creation might be truncated.\n");
++	}
+ 	if (*cmd && system(cmd) == -1) {
+ 		fprintf (stderr, "%s: system(%s) failed: %s\n",
+ 				params->cmdname, cmd, strerror(errno));
+diff -ur u-boot-2018.03.orig/tools/mkimage.h u-boot-2018.03/tools/mkimage.h
+--- u-boot-2018.03.orig/tools/mkimage.h	2019-10-24 16:07:20.643345838 +0200
++++ u-boot-2018.03/tools/mkimage.h	2019-10-24 17:27:38.682921690 +0200
+@@ -43,6 +43,6 @@
+ #define MKIMAGE_TMPFILE_SUFFIX		".tmp"
+ #define MKIMAGE_MAX_TMPFILE_LEN		256
+ #define MKIMAGE_DEFAULT_DTC_OPTIONS	"-I dts -O dtb -p 500"
+-#define MKIMAGE_MAX_DTC_CMDLINE_LEN	512
++#define MKIMAGE_MAX_DTC_CMDLINE_LEN	2 * MKIMAGE_MAX_TMPFILE_LEN + 35 /* 35 bytes for quotes, space, -o, len(MKIMAGE_DEFAULT_DTC_OPTIONS), len(MKIMAGE_TMPFILE_SUFFIX) */
+ 
+ #endif /* _MKIIMAGE_H_ */
